funding: implement reservation zombie sweeper

paddyquinn opened the pull request:
In this PR, a timeout is added to each `ChannelReservation`, which cancels the reservation when reached. On each successive message in the `ChannelReservation` workflow the timeout is reset, so that timeouts only occur after an idle period in the workflow.

This PR takes care of #60. Instead of creating a goroutine that periodically sweeps any zombie reservations, this PR adds a timer to each reservation, which cancels the reservation when it expires. In that way, a zombie reservation will be cleaned up (and thus resources freed up) as soon as possible.

paddyquinn commented:
@Roasbeef Thanks for the feedback. I just updated the PR. I moved the timeout logic to the `fundingManager` and now the timeout function calls `failFundingFlow` so that timeouts follow the same failure flow as other errors. Once a peer receives the last message before the funding transaction from its counterpart the timeout is stopped for good to ensure a reservation is not cancelled after the funding transaction has been broadcast. See below for reference.

Could also use some feedback on the length of `reservationTimeout` as I picked 1 minute pretty arbitrarily.

## Timeout Status in Different States of the Funding Workflow
**_handleInitFundingMsg_** - timeout is started on initiating peer after `lnwire.OpenChannel` is successfully sent to the remote peer

**_handleFundingOpen_** - timeout is started on remote peer once `lnwire.AcceptChannel` is successfully sent to the initiating peer

**_handleFundingAccept_** - timeout is stopped on initiating peer as soon as we get the reservation from the `fundingManager`'s map; timeout is restarted once `lnwire.FundingCreated` is successfully sent to the remote peer

**_handleFundingCreated_** - timeout is stopped for good on remote peer as soon as we get the reservation from the `fundingManager`'s map

**_handleFundingSigned_** - timeout is stopped for good on initiating peer as soon as we get the reservation from the `fundingManager`'s map

paddyquinn commented:
Realized that I was not explicitly tracking the state of the reservation as suggested and that was causing a possible race condition in which a funding message could be received for a reservation, but before the timeout for that reservation could be stopped in the processing of that funding message, the timeout expired causing us to fail that funding flow erroneously. Now, timeouts are also handled by the `reservationCoordinator` and state is tracked for each reservation. This solves that race condition by forcing the timeout to happen in the order it comes in to the `reservationCoordinator`. Each time a message is processed it updates the reservation's state. If a timeout message is received, `handleTimeoutMsg` ensures that the state has indeed not changed before proceeding to fail the funding flow.

halseth commented on fundingmanager.go:
```diff
@@ -53,6 +53,10 @@ const (
+
+	// reservationTimeout is the length of idle time that must pass before
+	// a reservation is cancelled
+	reservationTimeout = 1 * time.Minute
```
1) If you make this part of the `fundingConfig` then we can set it to a small value during tests.
2) 1 minute is a bit small maybe? I was thinking this should be on the order of ~10 minutes, but don't really have a convincing reason other than 1 minute being close to what we can expect for network latency+processing time on a slow network/computer.

paddyquinn replied:
Done.

halseth commented on fundingmanager.go:
```diff
@@ -753,6 +800,8 @@ func (f *fundingManager) reservationCoordinator() {
 			}
 		case req := <-f.fundingRequests:
 			f.handleInitFundingMsg(req)
+		case msg := <-f.timeoutMsgs:
```
I like this way of handling timeouts much better, but I think it can be simplified significantly.

To me it seems simpler to let each reservation have a `lastUpdated` timestamp. And instead of tracking the state of the funding message and modifying this in every case, I would make it possible to extract the reservation context from the messages in `fundingMsgs`, and each time a message is received, this timestamp can be updated. Now you can add a ticker to the reservation coordinator that at certain intervals loops through all active reservations, and cancels those that haven't been updated in a while. This makes sure the race condition you mentioned cannot occur, as they are never cancelled in parallel with handling messages.

paddyquinn replied:
I do not think the race condition can occur as it currently stands because, although each reservation's timer is running in parallel, when that timeout occurs it sends a timeout message to the reservation coordinator which handles messages it receives sequentially. Thus, if a timeout occurs after a new message is received for a reservation but before the reservation's timer can be stopped, the timeout will be ignored by `handleTimeoutMsg` because the states will be different. I chose this way of doing timeouts because then any stale reservation will get cleaned up right when it times out, and we do not have to run a "stop the world" zombie sweeper. However, I definitely see your point in that it makes the code more complex.

Anyway, I will write some tests to confirm that there is no race condition and then reduce the code complexity by moving to a zombie sweeper rather than individual timeouts, unless you think that the tradeoff in code complexity is worth avoiding a "stop the world" zombie sweeper.

Any suggestions on how often the zombie sweeper should run?

halseth replied:
That sounds like a good approach :)

With a 10 minute timeout I would say maybe let it run once a minute?

paddyquinn replied:
I updated the PR to move to a zombie sweeper implementation, rather than having individual timeouts. I still need to write tests but I wanted to get feedback on whether or not an error message should be sent to the peer when a reservation is timed out and cancelled. I assumed that we would want to send an error message so I included it. However, I left the message generic because of the following [comment in the code](https://github.com/lightningnetwork/lnd/blob/master/fundingmanager.go#L702-L703). Please advise on this. In the meantime, I will add tests.

halseth commented on fundingmanager.go:
```diff
@@ -1386,6 +1444,13 @@ func (f *fundingManager) handleFundingSigned(fmsg *fundingSignedMsg) {
+	// the timestamp to the zero value. This indicates to the zombie
+	// sweeper that this reservation should not be pruned, as it is now
+	// waiting for the funding transaction
+	resCtx.lastUpdated = time.Time{}
```
set it to a high future value instead?

halseth replied:
Actually, I think all the timestamp updates can be moved to the end of the corresponding methods (just use a `defer` just after getting the context?). This to ensure consistency, and because the timeout is long enough for it to not make any difference.

paddyquinn replied:
I like the idea of deferring the call. I will implement it that way and make it consistent across each step in the funding flow. Regarding setting it to a high future value, I think it is cleaner and more maintainable to set it to the time zero value. The time zero value is `January 1, year 1, 00:00:00 UTC` a time that will never be set by `time.Now`. If it is set to a future time, eventually this time would need to be changed. It also simplifies creating the time object (instantiate an empty `time.Time` struct, rather than having to specify the time down to the nanosecond via the `time.Date` function) and comparing the time object (just call `IsZero()`).

halseth replied:
Yeah, that's reasonable. I was thinking of something similar to Swift's `Date.distantFuture`, but doesn't look like Go has the equivalent :p

halseth commented on fundingmanager.go:
```diff
@@ -1189,6 +1240,13 @@ func (f *fundingManager) handleFundingCreated(fmsg *fundingCreatedMsg) {
+	// set the timestamp to the zero value. This indicates to the zombie
+	// sweeper that this reservation should not be pruned, as it is now
+	// waiting for the funding transaction
+	resCtx.lastUpdated = time.Time{}
```
do this first after the last message is sent to the peer.

halseth replied:
Also consider placing these timestamp changes into methods like
`updateLastUpdated(resCtx)` and `lockLastUpdated(resCtx)` or similar.

paddyquinn commented on fundingmanager.go:
```diff
@@ -2580,7 +2650,44 @@ func (f *fundingManager) handleErrorMsg(fmsg *fundingErrorMsg) {
+			if !resCtx.lastUpdated.IsZero() && time.Since(resCtx.lastUpdated) > f.cfg.ReservationTimeout {
+				errMsg := &lnwire.Error{
+					ChanID: pendingChanID,
+					Data:   lnwire.ErrorData("funding failed due to internal error"),
```
The reason we can't call `failFundingFlow` here is because a few lines above we acquire a write lock on the `resMtx` so that our access to the map is threadsafe. `failFundingFlow` eventually calls `deleteReservationCtx` which attempts to acquire a write lock on `resMtx`. Thus, if we called `failFundingFlow` within this loop we would end up in deadlock. Even if we switched the above lock to a read lock we would end up in deadlock because `deleteReservationCtx` requires a write lock.

halseth replied:
this can be fixed by fetching all the reservations to be deleted under the read lock, then calling `failFundingFlow` on them all after releasing it.

halseth commented on fundingmanager.go:
```diff
@@ -788,8 +796,9 @@ func (f *fundingManager) reservationCoordinator() {
-		case msg := <-f.pruneMsgs:
-			f.handlePruneMsg(msg)
+
+		case <-zombieSweepTicker.C:
```
👍

halseth commented on fundingmanager.go:
```diff
@@ -1241,11 +1230,10 @@ func (f *fundingManager) handleFundingCreated(fmsg *fundingCreatedMsg) {
+	// initiating peer before the funding transaction will be broadcast.
+	// Lock the reservation so it is not pruned by the zombie
+	// sweeper, as it is now waiting for the funding transaction.
+	resCtx.lock()
```
Do this towards the end of the method, right before launching the go routine. This to make it more clear that we only want to "lock" it from being swept in case we successfully start waiting for the funding confirmation.

paddyquinn replied:
done

halseth commented on fundingmanager.go:
```diff
@@ -1445,11 +1433,10 @@ func (f *fundingManager) handleFundingSigned(fmsg *fundingSignedMsg) {
+	// remote peer before we broadcast the funding transaction. Lock
+	// the reservation so it is not pruned by the zombie sweeper,
+	// as it is now waiting for the funding transaction.
+	resCtx.lock()
```
same

paddyquinn replied:
done

halseth commented on fundingmanager.go:
```diff
@@ -2507,6 +2494,9 @@ func (f *fundingManager) handleInitFundingMsg(msg *initFundingMsg) {
 	f.activeReservations[peerIDKey][chanID] = resCtx
 	f.resMtx.Unlock()
 
+	// Update the timestamp once the initFundingMsg has been handled
```
nit: missing period.

paddyquinn replied:
done

halseth commented on fundingmanager.go:
```diff
@@ -2650,41 +2636,28 @@ func (f *fundingManager) handleErrorMsg(fmsg *fundingErrorMsg) {
-	defer f.resMtx.Unlock()
+// pruneZombieReservations loops through all pending reservations and fails the
+// funding flow for any reservations that have not been updated since the
+// ReservationTimeout and are not locked waiting for the funding transaction
```
nit: period.

paddyquinn replied:
done

halseth commented on fundingmanager.go:
```diff
@@ -2650,41 +2636,28 @@ func (f *fundingManager) handleErrorMsg(fmsg *fundingErrorMsg) {
 	}
+	f.resMtx.RUnlock()
+
+	for pendingChanID, resCtx := range zombieReservations {
```
Excellent! 👍

halseth commented on fundingmanager.go:
```diff
@@ -77,6 +77,23 @@ type reservationWithCtx struct {
 }
 
+// isLocked checks the reservation's timestamp to determine whether it is locked.
+// The time zero value represents a locked reservation
```
nit: period. 

Also this last sentence probably can be moved into the method, as the caller doesn't need to know this.

paddyquinn replied:
done

halseth commented on fundingmanager.go:
```diff
@@ -77,6 +77,23 @@ type reservationWithCtx struct {
+}
+
+// lock locks the reservation from zombie pruning by setting its timestamp to the
+// zero value
```
nit: period.

paddyquinn replied:
done

halseth commented on fundingmanager.go:
```diff
@@ -77,6 +77,23 @@ type reservationWithCtx struct {
+	r.lastUpdated = time.Time{}
+}
+
+// updateTimestamp updates the reservation's timestamp with the current time
```
nit: period

paddyquinn replied:
done

halseth commented on fundingmanager.go:
```diff
@@ -977,6 +968,9 @@ func (f *fundingManager) handleFundingOpen(fmsg *fundingOpenMsg) {
 	f.activeReservations[peerIDKey][msg.PendingChannelID] = resCtx
 	f.resMtx.Unlock()
 
+	// Update the timestamp once the fundingOpenMsg has been handled
```
nit: period.

paddyquinn replied:
done

halseth commented on fundingmanager.go:
```diff
@@ -1089,6 +1079,9 @@ func (f *fundingManager) handleFundingAccept(fmsg *fundingAcceptMsg) {
 		return
 	}
 
+	// Update the timestamp once the fundingAcceptMsg has been handled
```
nit: period

paddyquinn replied:
done

paddyquinn commented:
@halseth Fixed the nits and squashed everything into 1 commit. I also pushed a new commit that added a check within `TestFundingManagerNormalWorkflow` to ensure that reservations are not pruned once they are locked. This test passes. However, I added a new test called `TestFundingManagerPeerTimeoutAfterInitFundingMsg` that is hanging on the call to `SendToPeer` within `pruneZombieReservations`. Can't quite figure out why. Need to do more digging. Wanted to push this up to get your feedback on the new tests and checks in the meantime, as I plan to add a new test for a timeout occurring after each funding message (`TestFundingManagerPeerTimeoutAfterFundingOpen`, `TestFundingManagerPeerTimeoutAfterFundingAccept`, `TestFundingManagerPeerTimeoutAfterFundingCreated`), all of which will follow a similar pattern.

halseth commented on fundingmanager.go:
```diff
@@ -263,6 +282,14 @@ type fundingConfig struct {
 	// sub-systems.
 	ReportShortChanID func(wire.OutPoint, lnwire.ShortChannelID) error
+
+	// ZombieSweeperInterval is the periodic time
```
nit: lines should be wrapped at 80 characters

paddyquinn replied:
Fixed

halseth commented on fundingmanager.go:
```diff
@@ -263,6 +282,14 @@ type fundingConfig struct {
+	// interval in which the zombie sweeper is run
+	ZombieSweeperInterval time.Duration
+
+	// ReservationTimeout is the length of idle time that must
```
same

paddyquinn replied:
Fixed

halseth commented on fundingmanager.go:
```diff
@@ -2580,7 +2637,31 @@ func (f *fundingManager) handleErrorMsg(fmsg *fundingErrorMsg) {
+	f.resMtx.RLock()
+	for _, pendingReservations := range f.activeReservations {
+		for pendingChanID, resCtx := range pendingReservations {
+			if !resCtx.isLocked() && time.Since(resCtx.lastUpdated) > f.cfg.ReservationTimeout {
```
to avoid this long line, you could do
```go
if resCtx.isLocked {
    continue
}

if time.Sinc.... {
```

Roasbeef replied:
👍

paddyquinn replied:
Done

halseth commented on lnwallet/reservation.go:
```diff
@@ -58,23 +58,23 @@ type InputScript struct {
 //
 // The reservation workflow consists of the following three steps:
-//  1. lnwallet.InitChannelReservation
+//  1. LightningWallet.InitChannelReservation
```
this should probably stay `lnwallet`

paddyquinn replied:
Fixed

halseth commented on lnwallet/reservation.go:
```diff
@@ -58,23 +58,23 @@ type InputScript struct {
 // ChannelReservation represents an intent to open a lightning payment channel
-// a counterparty. The funding processes from reservation to channel opening is
-// a 3-step process. In order to allow for full concurrency during the
+// with a counterparty. The funding processes from reservation to channel opening
```
Could you make the changes not in `fundingmanager` a separate commit labeled `lnwallet:`? :)

paddyquinn replied:
Done

halseth commented on fundingmanager_test.go:
```diff
@@ -579,6 +581,31 @@ func openChannel(t *testing.T, alice, bob *testNode, localFundingAmt,
+	}
+}
+
+func assertNumPendingReservations(t *testing.T, node *testNode, peerPubKey *btcec.PublicKey, expectedNum int) {
```
nit: long line

paddyquinn replied:
Fixed

halseth commented on fundingmanager_test.go:
```diff
@@ -304,8 +306,8 @@ func createTestFundingManager(t *testing.T, privKey *btcec.PrivateKey,
-		ZombieSweeperInterval: 1 * time.Minute,
-		ReservationTimeout:    10 * time.Minute,
+		ZombieSweeperInterval: 1 * time.Hour,
+		ReservationTimeout:    100 * time.Millisecond,
```
since you are controlling when the zombie sweeper run, you can make this as small as possible. This let us make the needed sleeps even tighter.

paddyquinn replied:
Done

halseth commented on fundingmanager_test.go:
```diff
@@ -1166,6 +1202,63 @@ func TestFundingManagerOfflinePeer(t *testing.T) {
+	assertErrorSent(t, alice)
+
+	// Alice's zombie reservation should have been pruned
+	assertNumPendingReservations(t, alice, bobPubKey, 0)
```
👍

paddyquinn commented:
@halseth Addressed your comments, fixed the tests, and added some new tests. Barring any other comments, this PR should be good to go now.

halseth commented on fundingmanager_test.go:
```diff
@@ -575,6 +581,32 @@ func openChannel(t *testing.T, alice, bob *testNode, localFundingAmt,
+
+func assertNumPendingReservations(t *testing.T, node *testNode,
+	peerPubKey *btcec.PublicKey, expectedNum int) {
+	numPendingReservations := len(node.fundingMgr.activeReservations[newSerializedKey(peerPubKey)])
```
nit: looks a bit long

paddyquinn replied:
Fixed.

halseth commented on fundingmanager_test.go:
```diff
@@ -1162,6 +1203,259 @@ func TestFundingManagerOfflinePeer(t *testing.T) {
 
+// TestFundingManagerPeerTimeoutAfterInitFunding checks that the zombie sweeper
+// will properly clean up a zombie reservation that times out after the
+// initFundingMsg has been handled
```
nit: period.

halseth commented on fundingmanager_test.go:
```diff
@@ -1162,6 +1203,259 @@ func TestFundingManagerOfflinePeer(t *testing.T) {
+	time.Sleep(1 * time.Millisecond)
+	go alice.fundingMgr.pruneZombieReservations()
+
+	// Alice should have sent an Error message to Bob
```
nit: period. Several more place in this file also.

paddyquinn replied:
Fixed.

halseth commented on fundingmanager_test.go:
```diff
@@ -1162,6 +1203,259 @@ func TestFundingManagerOfflinePeer(t *testing.T) {
+
+	_, ok = bobMsg.(*lnwire.AcceptChannel)
+	if !ok {
+		errorMsg, gotError := bobMsg.(*lnwire.Error)
```
This block of code is being used many places in this file now. Would be nice to extract into something like the `assertErrorSent` you added, just for generic messages:

```go
func assertSentMsg(msgChan chan *lnwire.Message, expected *lnwire.Message) (*lnwire.Message, error)
```

Not a blocker for this PR though.

paddyquinn replied:
Extracted this into `assertFundingMsgSent`. It was doing the same work as `checkNodeSendingFundingLocked` except in a more generic way so I removed `checkNodeSendingFundingLocked` and replaced any uses of it with `assertFundingMsgSent`.

halseth commented on fundingmanager_test.go:
```diff
@@ -868,7 +900,16 @@ func TestFundingManagerNormalWorkflow(t *testing.T) {
-	// Notify that transaction was mined
+	// Make sure both reservations time out and then run both zombie sweepers.
+	time.Sleep(1 * time.Millisecond)
+	alice.fundingMgr.pruneZombieReservations()
```
Note that in case they in face **do** prune the reservations here, this will block, making the test timeout instead of fail. Instead you should do something similar as you did in the other tests, checking that an error is _not_ sent:

```go
go pruneAlice()
go pruneBob()

select {
case <-aliceMsgChan:
      t.Fatalf("error sent unexpectedly")
case <- time.After(100 * time.Millisecond): // note that we use a bit higher timeout, to make sure the go routine spins up
       // expected
}

...same for bob...
```

paddyquinn replied:
Fixed.

cfromknecht commented on fundingmanager_test.go:
```diff
@@ -581,6 +542,81 @@ func openChannel(t *testing.T, alice, bob *testNode, localFundingAmt,
+	}
+}
+
+func assertFundingMsgSent(t *testing.T, msgChan chan lnwire.Message,
```
Nice work on this, definitely makes parsing the tests easier :)

halseth replied:
Yeh, looks great! 😀

cfromknecht commented on fundingmanager.go:
```diff
@@ -266,6 +285,14 @@ type fundingConfig struct {
+	ZombieSweeperInterval time.Duration
+
+	// ReservationTimeout is the length of idle time that must pass before a
+	// reservation is considered a zombie
```
nit: use complete sentences for comments!

paddyquinn replied:
Fixed.

cfromknecht commented on fundingmanager_test.go:
```diff
@@ -1131,7 +1163,8 @@ func TestFundingManagerOfflinePeer(t *testing.T) {
 	// This should make Alice send the fundingLocked.
-	fundingLockedAlice := checkNodeSendingFundingLocked(t, alice)
+	fundingLockedAlice :=
+		assertFundingMsgSent(t, alice.msgChan, "FundingLocked").(*lnwire.FundingLocked)
```
Really like how this turned out! I might suggest wrapping the lines like this:
```golang
fundingLockedAlice := assertFundingMsgSent(
        t, alice.msgChan, "FundingLocked",
).(*lnwire.FundingLocked)
```

feel free to do this on similar lines that are over 80 chars

paddyquinn replied:
Done. Pretty sure I got em all.

paddyquinn commented on fundingmanager.go:
```diff
@@ -2582,7 +2639,35 @@ func (f *fundingManager) handleErrorMsg(fmsg *fundingErrorMsg) {
+	f.resMtx.RUnlock()
+
+	for pendingChanID, resCtx := range zombieReservations {
+		err := fmt.Errorf("reservation timed out waiting for peer (peerID:%v, chanID:%x)",
```
Fixed.

halseth commented on fundingmanager_test.go:
```diff
@@ -1168,6 +1210,212 @@ func TestFundingManagerOfflinePeer(t *testing.T) {
+	alice.fundingMgr.initFundingWorkflow(bobAddr, initReq)
+
+	// Alice should have sent the OpenChannel message to Bob.
+	var aliceMsg lnwire.Message
```
Why not use `assertFundingMsgSent` here?

paddyquinn replied:
The reason I did not use `assertFundingMsgSent` here is because errors are handled differently in `initFundingWorkflow` so we must handle them differently here. Specifically, the `select` statement here as an additional `case` statement:

```
case err := <- initReq.err:
    t.Fatalf("error init funding workflow: %v", err)
```

halseth replied:
Makes sense, thanks!

halseth replied:
Can you check if your account is flagged by Travis? Seems like tests are not running. You might have to email them to get it unflagged.

halseth replied:
When that is done, you can rebase this on latest master :)

paddyquinn replied:
Not sure why my account would be flagged but I'll check with them, thanks!

paddyquinn commented:
@halseth This is the response I got from Travis CI.
> This is a part of our abuse prevention system, which blocks a repo if it looks like someone is trying to mine bitcoins. It's not a very sophisticated check, so often we have false positives. Could I ask how you're using Travis for this repository, and how you're using bitcoin in relation to Travis?

Just a heads up in case other people see the same issue.
